docs/hello_nf-core/04_input_validation.md

```diff
@@ -1,63 +1,616 @@
+                row[0]
+            }
+
+        emit:
+        samplesheet = ch_samplesheet
+        versions    = ch_versions
+    ```
+
+=== "Before"
+
+    ```groovy title="core-hello/subworkflows/local/utils_nfcore_hello_pipeline/main.nf" linenums="64"
+        //
+        // Create channel from input file provided through params.input
+        //
+        ch_samplesheet = Channel.fromPath(params.input)
+                            .splitCsv()
+                            .map { line -> line[0] }
+
+        emit:
+        samplesheet = ch_samplesheet
+        versions    = ch_versions
+    ```
+
+Let's break down what changed:
+
+1. **`samplesheetToList(params.input, "${projectDir}/assets/schema_input.json")`**: Validates the input file against our schema and returns a list
+2. **`Channel.fromList(...)`**: Converts the list into a Nextflow channel
+3. **`.map { row -> row[0] }`**: Extracts just the greeting string from each validated row (accessing the first column by index)
+
+!!! note "Parameter validation is enabled by default"
+
+    The pipeline template already includes parameter validation enabled via `params.validate_params = true` in `nextflow.config`. The validation is performed by the `UTILS_NFSCHEMA_PLUGIN` subworkflow during pipeline initialization.
 
 ### Takeaway
 
-You know how to [...].
+You've successfully implemented input data validation using `samplesheetToList` and JSON schemas.
 
 ### What's next?
 
-[...].
+Test both parameter and input data validation to see them in action.
+
+### 3.6. Test input validation
+
+Let's verify that our validation works by testing both valid and invalid inputs.
+
+**Test with valid input:**
+
+First, confirm the pipeline runs successfully with valid input:
+
+```bash
+nextflow run core-hello --outdir core-hello-results -profile test,docker
+```
+
+Note that we no longer need `--validate_params false` since validation is working!
+
+```console title="Output"
+ N E X T F L O W   ~  version 25.04.3
+
+Launching `./main.nf` [nasty_kalman] DSL2 - revision: c31b966b36
+
+Input/output options
+  input                     : /private/tmp/core-hello-test/assets/greetings.csv
+  batch                     : test
+  outdir                    : core-hello-results
+
+Institutional config options
+  config_profile_name       : Test profile
+  config_profile_description: Minimal test dataset to check pipeline function
+
+Core Nextflow options
+  runName                   : nasty_kalman
+  containerEngine           : docker
+  profile                   : test,docker
+
+!! Only displaying parameters that differ from the pipeline defaults !!
+------------------------------------------------------
+executor >  local (7)
+[cc/cc800d] CORE_HELLO:HELLO:sayHello (1)       | 3 of 3 ✔
+[d6/46ab71] CORE_HELLO:HELLO:convertToUpper (1) | 3 of 3 ✔
+[b2/3def99] CORE_HELLO:HELLO:CAT_CAT (test)     | 1 of 1 ✔
+[a3/f82e41] CORE_HELLO:HELLO:cowpy              | 1 of 1 ✔
+-[core/hello] Pipeline completed successfully-
+```
```

Comment:
I am getting the character warning: 

```
    ch_samplesheet = Channel.fromList(samplesheetToList(params.input, "${projectDir}/assets/schema_input.json"))
        .map { row ->
            // Extract just the greeting string from each row
            row[0]
        }
        ```
        
        TIL schema picks up any params not just once defined in the nextflow.config. However the schema builder is not as smart